# Practical Application III: Comparing Classifiers

**Overview**: In this practical application, your goal is to compare the performance of the classifiers we encountered in this section, namely K Nearest Neighbor, Logistic Regression, Decision Trees, and Support Vector Machines.  We will utilize a dataset related to marketing bank products over the telephone.  



### Getting Started

Our dataset comes from the UCI Machine Learning repository [link](https://archive.ics.uci.edu/ml/datasets/bank+marketing).  The data is from a Portugese banking institution and is a collection of the results of multiple marketing campaigns.  We will make use of the article accompanying the dataset [here](CRISP-DM-BANK.pdf) for more information on the data and features.



### Problem 1: Understanding the Data

To gain a better understanding of the data, please read the information provided in the UCI link above, and examine the **Materials and Methods** section of the paper.  How many marketing campaigns does this data represent?

In [1]:
#The dataset collected is related to 17 campaigns that occurred between May 2008 and November 2010,
#corresponding to a total of 79354 contacts.

### Problem 2: Read in the Data

Use pandas to read in the dataset `bank-additional-full.csv` and assign to a meaningful variable name.

In [2]:
import pandas as pd

In [3]:
df = pd.read_csv('data/bank-additional-full.csv', sep = ';')

In [4]:
# Set display options to show all columns
pd.set_option('display.max_columns', None)
df.sample(20)

age            job   marital            education  default housing  \
7097    58    blue-collar   married             basic.4y  unknown     yes   
36937   64     management   married    university.degree       no      no   
6172    43    blue-collar   married             basic.4y  unknown     yes   
8917    44     technician    single             basic.9y       no     yes   
4021    33         admin.    single          high.school  unknown     yes   
17006   47     technician  divorced          high.school       no     yes   
16151   58        retired   married             basic.4y       no     yes   
39093   76        retired  divorced  professional.course       no     yes   
33224   36         admin.   married          high.school       no     yes   
22845   59        retired   married             basic.9y       no     yes   
24198   39  self-employed   married    university.degree       no      no   
32952   29    blue-collar   married             basic.9y       no     yes   
6392    40    blue-collar   married          high.school  unknown      no   
7708    31         admin.   married    university.degree       no      no   
37549   48     management  divorced    university.degree       no      no   
7323    59    blue-collar   married             basic.4y  unknown      no   
30194   48   entrepreneur   married    university.degree       no      no   
13021   28       services   married          high.school       no     yes   
1581    30         admin.   married          high.school       no     yes   
13062   28       services   married          high.school       no     yes   

      loan    contact month day_of_week  duration  campaign  pdays  previous  \
7097    no  telephone   may         thu       101         1    999         0   
36937   no   cellular   jun         tue       205         1    999         0   
6172    no  telephone   may         tue       193         1    999         0   
8917    no  telephone   jun         thu       113         3    999         0   
4021    no  telephone   may         mon        68         2    999         0   
17006   no   cellular   jul         thu       828         2    999         0   
16151   no   cellular   jul         tue       614         1    999         0   
39093   no   cellular   dec         thu       956         1    999         0   
33224   no   cellular   may         tue       300         5    999         0   
22845   no   cellular   aug         mon        93         6    999         0   
24198   no   cellular   nov         mon        85         1    999         1   
32952  yes   cellular   may         mon        34         1    999         1   
6392    no  telephone   may         tue        43         5    999         0   
7708    no  telephone   may         fri       370         3    999         0   
37549   no   cellular   aug         fri       117         1    999         0   
7323    no  telephone   may         thu       763         2    999         0   
30194   no   cellular   apr         thu       294         2      2         1   
13021   no   cellular   jul         tue       286         1    999         0   
1581    no  telephone   may         fri       110         4    999         0   
13062  yes  telephone   jul         tue       153         1    999         0   

          poutcome  emp.var.rate  cons.price.idx  cons.conf.idx  euribor3m  \
7097   nonexistent           1.1          93.994          -36.4      4.860   
36937  nonexistent          -2.9          92.963          -40.8      1.099   
6172   nonexistent           1.1          93.994          -36.4      4.857   
8917   nonexistent           1.4          94.465          -41.8      4.866   
4021   nonexistent           1.1          93.994          -36.4      4.858   
17006  nonexistent           1.4          93.918          -42.7      4.962   
16151  nonexistent           1.4          93.918          -42.7      4.961   
39093  nonexistent          -3.0          92.713          -33.0      0.707   
33224 

### Problem 3: Understanding the Features


Examine the data description below, and determine if any of the features are missing values or need to be coerced to a different data type.


```
Input variables:
# bank client data:
1 - age (numeric)
2 - job : type of job (categorical: 'admin.','blue-collar','entrepreneur','housemaid','management','retired','self-employed','services','student','technician','unemployed','unknown')
3 - marital : marital status (categorical: 'divorced','married','single','unknown'; note: 'divorced' means divorced or widowed)
4 - education (categorical: 'basic.4y','basic.6y','basic.9y','high.school','illiterate','professional.course','university.degree','unknown')
5 - default: has credit in default? (categorical: 'no','yes','unknown')
6 - housing: has housing loan? (categorical: 'no','yes','unknown')
7 - loan: has personal loan? (categorical: 'no','yes','unknown')
# related with the last contact of the current campaign:
8 - contact: contact communication type (categorical: 'cellular','telephone')
9 - month: last contact month of year (categorical: 'jan', 'feb', 'mar', ..., 'nov', 'dec')
10 - day_of_week: last contact day of the week (categorical: 'mon','tue','wed','thu','fri')
11 - duration: last contact duration, in seconds (numeric). Important note: this attribute highly affects the output target (e.g., if duration=0 then y='no'). Yet, the duration is not known before a call is performed. Also, after the end of the call y is obviously known. Thus, this input should only be included for benchmark purposes and should be discarded if the intention is to have a realistic predictive model.
# other attributes:
12 - campaign: number of contacts performed during this campaign and for this client (numeric, includes last contact)
13 - pdays: number of days that passed by after the client was last contacted from a previous campaign (numeric; 999 means client was not previously contacted)
14 - previous: number of contacts performed before this campaign and for this client (numeric)
15 - poutcome: outcome of the previous marketing campaign (categorical: 'failure','nonexistent','success')
# social and economic context attributes
16 - emp.var.rate: employment variation rate - quarterly indicator (numeric)
17 - cons.price.idx: consumer price index - monthly indicator (numeric)
18 - cons.conf.idx: consumer confidence index - monthly indicator (numeric)
19 - euribor3m: euribor 3 month rate - daily indicator (numeric)
20 - nr.employed: number of employees - quarterly indicator (numeric)

Output variable (desired target):
21 - y - has the client subscribed a term deposit? (binary: 'yes','no')
```



In [5]:
missing_values = df.isnull().sum()
print(missing_values)

age               0
job               0
marital           0
education         0
default           0
housing           0
loan              0
contact           0
month             0
day_of_week       0
duration          0
campaign          0
pdays             0
previous          0
poutcome          0
emp.var.rate      0
cons.price.idx    0
cons.conf.idx     0
euribor3m         0
nr.employed       0
y                 0
dtype: int64


## See problem 5 on how I applied one-hot coding to transform the columns from categorical to numeric features

### Problem 4: Understanding the Task

After examining the description and data, your goal now is to clearly state the *Business Objective* of the task.  State the objective below.

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 41188 entries, 0 to 41187
Data columns (total 21 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   age             41188 non-null  int64  
 1   job             41188 non-null  object 
 2   marital         41188 non-null  object 
 3   education       41188 non-null  object 
 4   default         41188 non-null  object 
 5   housing         41188 non-null  object 
 6   loan            41188 non-null  object 
 7   contact         41188 non-null  object 
 8   month           41188 non-null  object 
 9   day_of_week     41188 non-null  object 
 10  duration        41188 non-null  int64  
 11  campaign        41188 non-null  int64  
 12  pdays           41188 non-null  int64  
 13  previous        41188 non-null  int64  
 14  poutcome        41188 non-null  object 
 15  emp.var.rate    41188 non-null  float64
 16  cons.price.idx  41188 non-null  float64
 17  cons.conf.idx   41188 non-null 

## Business Objective: This task is to choose the right features that are critical and choose the best machine learning model to predict with highest accuracy result which banks can optimize their strategies to attract more customers to sign up for long-term deposit subscriptions.

### Problem 5: Engineering Features

Now that you understand your business objective, we will build a basic model to get started.  Before we can do this, we must work to encode the data.  Using just the bank information features, prepare the features and target column for modeling with appropriate encoding and transformations.

In [7]:
# Drop the label column 'y' and maintain all other columns
X = df.drop(columns='y')
y = df['y']

In [8]:
from sklearn.compose import make_column_transformer
from sklearn.preprocessing import OneHotEncoder, StandardScaler
# Define column transformer for handling different types of encodings
transformer = make_column_transformer((OneHotEncoder(drop = 'if_binary'), ['default','job','marital','education','housing','loan','contact','month','day_of_week','poutcome']), 
                                     remainder = StandardScaler())

### Problem 6: Train/Test Split

With your data prepared, split it into a train and test set.

In [9]:
from sklearn.model_selection import train_test_split
# Split the dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=42)

### Problem 7: A Baseline Model

Before we build our first model, we want to establish a baseline.  What is the baseline performance that our classifier should aim to beat?

## My expectation for the baseline performanced of a classifier should be greater than 80% (or 0.8000) accuracy. 

### Problem 8: A Simple Model

Use Logistic Regression to build a basic model on your data.  

In [10]:
from sklearn.pipeline import Pipeline
# Create a pipeline with the transformer and and the classifier 'Logistic Regression'
from sklearn.linear_model import LogisticRegression
logregressionpipe = Pipeline([
    ('transform', transformer),
    ('classifier', LogisticRegression(max_iter=1000))
])
logregressionpipe.fit(X_train, y_train)

Pipeline(steps=[('transform',
                 ColumnTransformer(remainder=StandardScaler(),
                                   transformers=[('onehotencoder',
                                                  OneHotEncoder(drop='if_binary'),
                                                  ['default', 'job', 'marital',
                                                   'education', 'housing',
                                                   'loan', 'contact', 'month',
                                                   'day_of_week',
                                                   'poutcome'])])),
                ('classifier', LogisticRegression(max_iter=1000))])

### Problem 9: Score the Model

What is the accuracy of your model?

In [11]:
#Calculate and print the training accuracy
logregressionpipe_acc_train = logregressionpipe.score(X_train, y_train)
logregressionpipe_acc_test = logregressionpipe.score(X_test, y_test)

In [12]:
print(f"Training Accuracy: {logregressionpipe_acc_train:.4f}")
print(f"Testing Accuracy: {logregressionpipe_acc_test:.4f}")

Training Accuracy: 0.9115
Testing Accuracy: 0.9128


### Problem 10: Model Comparisons

Now, we aim to compare the performance of the Logistic Regression model to our KNN algorithm, Decision Tree, and SVM models.  Using the default settings for each of the models, fit and score each.  Also, be sure to compare the fit time of each of the models.  Present your findings in a `DataFrame` similar to that below:

| Model | Train Time | Train Accuracy | Test Accuracy |
| ----- | ---------- | -------------  | -----------   |
|     |    |.     |.     |

In [13]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score
from sklearn.svm import SVC
import time

In [14]:
# Convert target variable to binary (yes=1, no=0)
y = y.map({'yes': 1, 'no': 0})

In [15]:
# Define the models to test (with baseline parameters)

models = {
    'KNN': KNeighborsClassifier(n_neighbors=5),
    'Logistic Regression': LogisticRegression(max_iter=1000),
    'Decision Tree': DecisionTreeClassifier(random_state=42),
    'SVC': SVC()
}

In [16]:
# Initialize a dictionary to store the accuracy results
model_performance = {}

In [17]:
# Fit and evaluate each model, and create a pipeline with the transformer and and the classifiers
for model_name, model in models.items():
    print(f"Processing model: {model_name}")
    
    pipe = Pipeline([
    ('transform', transformer),
    ('models', model)
    ])
    
    # Track the start time for training
    start_time = time.time()
    
    # Fit the pipeline
    pipe.fit(X_train, y_train)
    
    # Calculate training time
    train_time = time.time() - start_time
    
    # Calculate train accuracy
    train_accuracy = accuracy_score(y_train, pipe.predict(X_train))
    
    # Evaluate the model on the test data (test accuracy)
    test_accuracy = accuracy_score(y_test, pipe.predict(X_test))
    
    # Store the results in the dictionary
    model_performance[model_name] = {
        'Train Time (s)': train_time,
        'Train Accuracy': train_accuracy,
        'Test Accuracy': test_accuracy
    }
    
    # Print the results for each model after processing
    print(f"{model_name} - Train Time: {train_time:.4f} seconds, Train Accuracy: {train_accuracy:.4f}, Test Accuracy: {test_accuracy:.4f}")

Processing model: KNN
KNN - Train Time: 0.1719 seconds, Train Accuracy: 0.9271, Test Accuracy: 0.9032
Processing model: Logistic Regression
Logistic Regression - Train Time: 0.9554 seconds, Train Accuracy: 0.9115, Test Accuracy: 0.9128
Processing model: Decision Tree
Decision Tree - Train Time: 0.4816 seconds, Train Accuracy: 1.0000, Test Accuracy: 0.8864
Processing model: SVC
SVC - Train Time: 29.2499 seconds, Train Accuracy: 0.9227, Test Accuracy: 0.9128


In [18]:
import pandas as pd
performance_df = pd.DataFrame(model_performance).T

In [19]:
# Rename the columns to match the desired output format
performance_df.columns = ['Train Time (s)', 'Train Accuracy', 'Test Accuracy']

# Display the DataFrame
print(performance_df)

                     Train Time (s)  Train Accuracy  Test Accuracy
KNN                        0.171949        0.927099       0.903176
Logistic Regression        0.955351        0.911463       0.912790
Decision Tree              0.481647        1.000000       0.886375
SVC                       29.249944        0.922696       0.912790


## Based on the result, either Logistic Regression or SVC yield the best test accuracy, however, with train time being a main factor of how efficient a model run, I would choose Logistic Regression over SVC. 

### Problem 11: Improving the Model

Now that we have some basic models on the board, we want to try to improve these.  Below, we list a few things to explore in this pursuit.

- More feature engineering and exploration.  For example, should we keep the gender feature?  Why or why not?
- Hyperparameter tuning and grid search.  All of our models have additional hyperparameters to tune and explore.  For example the number of neighbors in KNN or the maximum depth of a Decision Tree.  
- Adjust your performance metric

### There is no gender feature for this dataset. Based on the article there are a few features that are more important than the others. 

## I try to improve KNN with grid search to see if it can perform better than Logistic Regression/SVC

In [20]:
knn_pipe = Pipeline([('transform', transformer), ('knn', KNeighborsClassifier(n_neighbors = 10))])

In [21]:
knn_pipe.fit(X_train, y_train)

Pipeline(steps=[('transform',
                 ColumnTransformer(remainder=StandardScaler(),
                                   transformers=[('onehotencoder',
                                                  OneHotEncoder(drop='if_binary'),
                                                  ['default', 'job', 'marital',
                                                   'education', 'housing',
                                                   'loan', 'contact', 'month',
                                                   'day_of_week',
                                                   'poutcome'])])),
                ('knn', KNeighborsClassifier(n_neighbors=10))])

In [22]:
knn_pipe.score(X_test, y_test)

0.9078372341458677

In [23]:
y_pred = knn_pipe.predict(X_test)

In [24]:
from sklearn.metrics import classification_report

In [25]:
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

          no       0.92      0.98      0.95      9144
         yes       0.66      0.36      0.47      1153

    accuracy                           0.91     10297
   macro avg       0.79      0.67      0.71     10297
weighted avg       0.89      0.91      0.90     10297



In [26]:
import numpy as np
# Set up the parameter grid (low range from 1 to 11 due to computational expensive for current data set)
#for the hyperparameter K
param_grid = {
    'knn__n_neighbors': np.arange(1, 11),
    'knn__weights': ['uniform', 'distance'],
    'knn__p': [1, 2]
}

In [27]:
from sklearn.model_selection import GridSearchCV
grid_search = GridSearchCV(knn_pipe, param_grid, cv=5, scoring='accuracy')


In [28]:
# Limit Data Sampling Size due to computationally expensive and time-consuming

X_train_sample = X_train.sample(frac=0.1, random_state=42)  # Use 10% of the data
y_train_sample = y_train.loc[X_train_sample.index]

In [29]:
# Fit grid search on the sample data (Use 10% of the data)
grid_search.fit(X_train_sample, y_train_sample)

GridSearchCV(cv=5,
             estimator=Pipeline(steps=[('transform',
                                        ColumnTransformer(remainder=StandardScaler(),
                                                          transformers=[('onehotencoder',
                                                                         OneHotEncoder(drop='if_binary'),
                                                                         ['default',
                                                                          'job',
                                                                          'marital',
                                                                          'education',
                                                                          'housing',
                                                                          'loan',
                                                                          'contact',
                                                        

In [31]:
# Best K value
best_k = grid_search.best_params_['knn__n_neighbors']
print(f"The best K value: {best_k}")

The best K value: 10


In [32]:
# Best accuracy
best_score = grid_search.best_score_
print(f"Best cross-validated accuracy: {best_score}")

Best cross-validated accuracy: 0.9099982691067018


### For banks attempting to use ML model to classify whether customers will sign up for long-term subscriptions and deposit money, I evaluate the consequence of false positive and false negatives. 

1. False Positive - model predicts that a customer will deposit money, but in reality, they do not. The consequence is that the bank may spend resources on customers who ultimately do not subscribe, and this leads to wasted resources and potential opportunity costs.

2. False Negative - model predicts that a customer will not deposit money, but in reality, they do. The consequence is that the bank misses out great opportunities to follow up with potential customers who would have subscribed. This could result in lost revenuse and sales opportunities. 

3. False negative is more detirmental in this situation, if bank fails to realize potential customers, the bank will miss out significant revenue opportunites. Although false positives lead to wasted resources, they do not directly result in lost sales, just inefficiency.

4. Avoiding false negatives is crucial because missing out on potential deposits (subscribing customers) can have a larger impact on the bank's profitability.

5. Recall is the metric to use, since it measure the proportion of actual positives (customers who subscribed) that were correctly identified by the model. Maximizing recall ensures that most potential subscribers are correctly identified, reducing the chances of missing out on revenue.

6. For my KNN modeling, the recall for "yes" is only 0.36. It is recommmended for further feature engineering and transformations, hyperparameters tuning to further improve this value, which ultimately could provide better prediction and insight on how to follow up with potential customers. 




##### Questions